factory/src/main/kotlin/io/qalipsis/core/factory/meters/CampaignMeterRegistryImpl.kt

```diff
@@ -19,146 +19,181 @@
+import io.qalipsis.api.meters.MeterType
 import io.qalipsis.api.meters.Timer
 import io.qalipsis.core.factory.campaign.Campaign
 import io.qalipsis.core.factory.campaign.CampaignLifeCycleAware
 import io.qalipsis.core.factory.configuration.FactoryConfiguration
-import io.qalipsis.core.meters.NoopMeterRegistry
 import io.qalipsis.core.reporter.MeterReporter
+import jakarta.inject.Named
 import jakarta.inject.Singleton
 import java.time.Duration
-import java.util.concurrent.ConcurrentHashMap
+import java.time.Instant
+import java.util.concurrent.CountDownLatch
 import java.util.function.ToDoubleFunction
-import io.micrometer.core.instrument.DistributionSummary as MicrometerDistributionSummary
-import io.micrometer.core.instrument.Gauge as MicrometerGauge
-import io.micrometer.core.instrument.Timer as MicrometerTimer
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.launch
+import kotlin.concurrent.fixedRateTimer
 
 /**
  * Implementation of meter registry.
  */
 @Singleton
 internal class CampaignMeterRegistryImpl(
-    private val factories: Collection<MeterRegistryFactory>,
+    private val factories: Collection<MeasurementPublisherFactory>,
     private val factoryConfiguration: FactoryConfiguration,
     private val meterReporter: MeterReporter,
-    @Value("\${${MetersConfig.EXPORT_CONFIGURATION}.campaign-step:PT5S}") private val step: Duration
+    @Value("\${${MetersConfig.EXPORT_CONFIGURATION}.campaign-step:PT5S}") private val step: Duration,
+    @Value("\${${MetersConfig.EXPORT_CONFIGURATION}.percentile:25.0, 50.0, 75.0, 99.0}") private val percentiles: List<Double>,
+    @Value("\${${MetersConfig.EXPORT_CONFIGURATION}.histogram-count:0}") private val histogramCount: List<Long>,
+    @Named(Executors.BACKGROUND_EXECUTOR_NAME) private val coroutineScope: CoroutineScope,
 ) : CampaignMeterRegistry, CampaignLifeCycleAware {
 
-    private var meterRegistry: MeterRegistry = NoopMeterRegistry()
+    private val additionalTags = mutableMapOf<String, String>()
 
-    private val additionalTags = mutableListOf<Tag>()
+    private lateinit var currentCampaignKey: CampaignKey
 
-    private val clock = Clock.SYSTEM
+    private var publishingJob: Job? = null
 
-    private lateinit var currentCampaignKey: CampaignKey
+    private var snapshots = listOf<MeterSnapshot<*>>()
+
+    private var publishers = listOf<MeasurementPublisher>()
 
     /**
      * Contains all the unique meters.
      */
-    private val meters = ConcurrentHashMap<Meter.Id, Meter<*>>()
+    val meters = HashMap<Meter.Id, Meter<*>>()
 
     override suspend fun init(campaign: Campaign) {
         meters.clear()
 
         currentCampaignKey = campaign.campaignKey
         if (factories.isNotEmpty()) {
             // Additional tags to force to all the created meters.
+            additionalTags.putAll(factoryConfiguration.tags)
             factoryConfiguration.zone?.takeIf { it.isNotBlank() }?.let { zone ->
```

Comment:
Move the initialization of the additionalTags into an init{} block to do it only once. the additionalTags are not campaign dependent.